File: scripts/spatial_b1/naming.py
from __future__ import annotations

import re
from typing import Any

ARABIC_PATTERN = re.compile(r"[\u0600-\u06ff]")
LATIN_PATTERN = re.compile(r"[A-Za-z]")


def _text(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    normalized = value.strip()
    return normalized or None


def _english_text(value: Any) -> str | None:
    candidate = _text(value)
    if not candidate or ARABIC_PATTERN.search(candidate) or not LATIN_PATTERN.search(candidate):
        return None
    return candidate


def _collect_strings(value: Any) -> list[str]:
    if isinstance(value, str):
        return [value.strip()] if value.strip() else []
    if isinstance(value, list):
        return [item for child in value for item in _collect_strings(child)]
    if isinstance(value, dict):
        return [item for child in value.values() for item in _collect_strings(child)]
    return []


def normalized_names(properties: dict[str, Any], provider: str, category: str, provider_id: str) -> dict[str, Any]:
    names = properties.get("names") if isinstance(properties.get("names"), dict) else {}
    common = names.get("common") if isinstance(names.get("common"), dict) else {}
    primary = names.get("primary")
    structured_primary_english = None
    if isinstance(primary, dict):
        structured_primary_english = _text(primary.get("en"))
        if structured_primary_english is None and str(primary.get("language") or "").lower().startswith("en"):
            structured_primary_english = _text(primary.get("value"))
    else:
        structured_primary_english = _english_text(primary)

    structured_common_english = _text(common.get("en"))
    osm_english = _text(properties.get("name:en")) if provider == "osm" else None
    name_en = _text(properties.get("name_en"))
    local_name = _text(properties.get("name"))
    english_name = next(
        (
            candidate
            for candidate in (structured_primary_english, structured_common_english, osm_english, name_en)
            if candidate
        ),
        None,
    )
    source_object_name = local_name or _text(primary) or english_name
    fallback = f"{category.replace('_', ' ').title()} — {provider}:{provider_id}"
    display_name = english_name or local_name or fallback
    alternate_candidates = [
        *_collect_strings(names.get("alternate")),
        *_collect_strings(common),
        _text(properties.get("alt_name")),
        _text(properties.get("official_name")),
        _text(properties.get("short_name")),
        source_object_name,
    ]
    alternate_names = sorted(
        {
            candidate
            for candidate in alternate_candidates
            if candidate and candidate not in {display_name, local_name, english_name}
        }
    )
    return {
        "displayName": display_name,
        "sourceObjectName": source_object_name,
        "localName": local_name,
        "englishName": english_name,
        "alternateNames": alternate_names,
        "displayNameSource": next(
            label
            for label, candidate in (
                ("structured_provider_primary_english", structured_primary_english),
                ("structured_provider_common_english", structured_common_english),
                ("osm_name_en", osm_english),
                ("name_en", name_en),
                ("local_name", local_name),
                ("neutral_category_provider_identity", fallback),
            )
            if candidate == display_name
        ),
    }

File: src/lib/spatial-b1/naming.ts
type NameRecord = Record<string, unknown>;

function recordValue(value: unknown): NameRecord {
  return value && typeof value === "object" && !Array.isArray(value) ? (value as NameRecord) : {};
}

function textValue(value: unknown) {
  return typeof value === "string" && value.trim() ? value.trim() : null;
}

function englishText(value: unknown) {
  const candidate = textValue(value);
  return candidate && /[A-Za-z]/.test(candidate) && !/[\u0600-\u06ff]/.test(candidate) ? candidate : null;
}

function collectStrings(value: unknown): string[] {
  if (typeof value === "string") return value.trim() ? [value.trim()] : [];
  if (Array.isArray(value)) return value.flatMap(collectStrings);
  if (value && typeof value === "object") return Object.values(value as NameRecord).flatMap(collectStrings);
  return [];
}

export function normalizeSpatialNamesV1(input: {
  properties: NameRecord;
  provider: "osm" | "overture";
  category: string;
  providerId: string;
}) {
  const names = recordValue(input.properties.names);
  const common = recordValue(names.common);
  const primaryRecord = recordValue(names.primary);
  const primaryLanguage = textValue(primaryRecord.language)?.toLowerCase();
  const structuredPrimaryEnglish =
    textValue(primaryRecord.en) ??
    (primaryLanguage?.startsWith("en") ? textValue(primaryRecord.value) : null) ??
    englishText(names.primary);
  const structuredCommonEnglish = textValue(common.en);
  const osmEnglish = input.provider === "osm" ? textValue(input.properties["name:en"]) : null;
  const nameEn = textValue(input.properties.name_en);
  const localName = textValue(input.properties.name);
  const englishName = structuredPrimaryEnglish ?? structuredCommonEnglish ?? osmEnglish ?? nameEn;
  const sourceObjectName = localName ?? textValue(names.primary) ?? englishName;
  const fallback = `${input.category.replace(/_/g, " ").replace(/\b\w/g, (value) => value.toUpperCase())} — ${input.provider}:${input.providerId}`;
  const displayName = englishName ?? localName ?? fallback;
  const alternateNames = [...new Set([
    ...collectStrings(names.alternate),
    ...collectStrings(common),
    textValue(input.properties.alt_name),
    textValue(input.properties.official_name),
    textValue(input.properties.short_name),
    sourceObjectName
  ].filter((value): value is string => Boolean(value) && ![displayName, localName, englishName].includes(value)))].sort();
  const displayNameSource =
    displayName === structuredPrimaryEnglish
      ? "structured_provider_primary_english"
      : displayName === structuredCommonEnglish
        ? "structured_provider_common_english"
        : displayName === osmEnglish
          ? "osm_name_en"
          : displayName === nameEn
            ? "name_en"
            : displayName === localName
              ? "local_name"
              : "neutral_category_provider_identity";

  return {
    displayName,
    sourceObjectName,
    localName,
    englishName,
    alternateNames,
    displayNameSource
  };
}
